environments/builder/builder.go

```diff
@@ -0,0 +1,154 @@
+package builder
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/mholt/archiver"
+
+	"github.com/fission/fission"
+)
+
+const (
+	pathPrefix      = "/tmp/fission"
+	srcPkgPath      = pathPrefix + "/srcPkg"
+	deployPkgPath   = pathPrefix + "/deployPkg"
+	archiveFilename = "archive.zip"
+
+	// supported environment variables
+	envSrcPkg    = "SRC_PKG"
+	envDeployPkg = "DEPLOY_PKG"
+)
+
+type (
+	Builder struct {
+		sharedVolumePath string
+	}
+)
+
+func MakeBuilder(sharedVolumePath string) *Builder {
+	return &Builder{
+		sharedVolumePath: sharedVolumePath,
+	}
+}
+
+func (builder *Builder) Handler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "", 404)
+		return
+	}
+
+	startTime := time.Now()
+	defer func() {
+		elapsed := time.Now().Sub(startTime)
+		log.Printf("elapsed time in build request = %v", elapsed)
+	}()
+
+	// parse request
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("Error reading request body")
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	var req fission.PackageBuildRequest
+	err = json.Unmarshal(body, &req)
+	if err != nil {
+		log.Printf("Error reading request body: %v", err)
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	log.Printf("builder received request: %v", req)
+
+	srcPkgArchivePath := filepath.Join(builder.sharedVolumePath, req.SrcPkgFilename)
+	err = builder.unarchive(srcPkgArchivePath, srcPkgPath)
+	if err != nil {
+		e := errors.New(fmt.Sprintf("Failed to unarchive source package: %v", err))
+		http.Error(w, e.Error(), 500)
+		return
+	}
+
+	buildLogs, err := builder.build(req.BuildCommand, srcPkgPath)
+	if err != nil {
+		e := errors.New(fmt.Sprintf("Failed to build source package: %v", err))
+		http.Error(w, e.Error(), 500)
+		return
+	}
+
+	archivePath := filepath.Join(builder.sharedVolumePath, archiveFilename)
+	err = builder.archive(deployPkgPath, archivePath)
+	if err != nil {
+		e := errors.New(fmt.Sprintf("Failed to archive deployment package: %v", err))
+		http.Error(w, e.Error(), 500)
+		return
+	}
+
+	resp := fission.PackageBuildResponse{
+		ArchiveFilename: archiveFilename,
+		BuildLogs:       buildLogs,
+	}
+
+	rBody, err := json.Marshal(resp)
+	if err != nil {
+		e := errors.New(fmt.Sprintf("Failed to encode response body: %v", err))
+		http.Error(w, e.Error(), 500)
+		return
+	}
+
+	w.Header().Add("Content-Type", "application/json")
+	w.Write(rBody)
+	w.WriteHeader(http.StatusOK)
+}
+
+func (builder *Builder) build(command string, workDir string) (string, error) {
+	// use `/bin/sh -c` to run multiple commands at the same time
+	cmdStrs := []string{"-c", command}
+	cmd := exec.Command("/bin/sh", cmdStrs...)
```

Comment:
So... is there a way to stream output to stdout while this is running?

It would be nice to be able to look at build output "live", by using `kubectl logs`, while the build is in progress.

Reply:
(Feel free to address this in a later change. You can probably do it with some complicated usage of the `exec` package.)

Reply:
Check out the binary environment for ways to work with the stdin and stdout of the forked process.

Reply:
@soamvasani Thinking about it, this would be quite some effort to implement, as currently the builder is implemented in a synchronous way (if i understand @life1347 explanation correctly). It would be a nice to have, but that would require the client-builder communication to be asynchronous, in which the client would be polling the server for updates on a build.

Reply:
I'm not talking about a change to the builder API. Just saying that it would be useful if the output were logged to stdout while the command is running.

Reply:
Yeah. I'm trying to avoid having too much in one PR. Let's talk about a progress reporting API in a different issue/PR.